tests/cpp/test_combined_inner_outer_reduction.cpp

```diff
@@ -1385,4 +1385,161 @@ TEST(StaticWarpReductionTest, StaticWarpReductionValidation) {
+
+  // Generate expected outputs using ATen computations
+  auto t0_fp32 = t0.to(at::kFloat);
+  auto t2_fp32 = t2.to(at::kFloat);
+  auto t3_fp32 = t3.to(at::kFloat);
+  auto t4_fp32 = t4.to(at::kFloat);
+
+  // Step-by-step computation matching the fusion
+  auto tv8 = t0_fp32.unsqueeze(0).expand({dim0, dim1}); // broadcast t0
+  auto tv12 = t1.unsqueeze(1).expand({dim0, 1}); // broadcast t1
+  auto tv13 = t2_fp32; // cast t2 to float
+  auto tv14 = tv8; // cast tv8 to float
+  auto tv18 = tv12.expand({dim0, dim1}); // expand tv12
+  auto tv19 = t3_fp32; // cast t3 to float
+  auto tv20 = tv14 * tv13; // mul(tv14, tv13)
+  auto tv21 = tv19 - tv18; // sub(tv19, tv18)
+  auto tv22 = tv21 * tv20; // mul(tv21, tv20)
+  auto tv23 = tv22.sum(1, false); // sum(tv22, {1})
+  auto tv27 = tv23.unsqueeze(1).expand({dim0, 1}); // broadcast tv23
+  auto tv31 = t4_fp32.expand({dim0, dim1}); // expand t4
+  auto tv33 = t4_fp32.pow(3.0); // pow(t4, 3.0)
+  auto tv35 = -0.5 * tv27; // mul(-0.5, tv27)
+  auto tv36 = tv31 * tv20; // mul(tv31, tv20)
+  auto tv37 = tv35 * tv33; // mul(tv35, tv33)
+  auto tv38 = -tv36; // neg(tv36)
+  auto tv39 = tv37.sum(1, false); // sum(tv37, {1})
+  auto tv40 = tv38.sum(1, false); // sum(tv38, {1})
+  auto tv44 = t1.unsqueeze(1).expand({dim0, 1}); // broadcast t1
+  auto tv48 = tv39.unsqueeze(1).expand({dim0, 1}); // broadcast tv39
+  auto tv52 = tv40.unsqueeze(1).expand({dim0, 1}); // broadcast tv40
+  auto tv56 = tv44.expand({dim0, dim1}); // expand tv44
+  auto tv60 = tv48.expand({dim0, dim1}); // expand tv48
+  auto tv61 = tv52.sum(1, false); // sum(tv52, {1})
+  auto tv62 = tv19 - tv56; // sub(tv19, tv56)
+  auto tv64 = 2.0 * tv60; // mul(2.0, tv60)
+  auto tv68 = tv61.unsqueeze(1).expand({dim0, 1}); // broadcast tv61
+  auto tv69 = tv64 * tv62; // mul(tv64, tv62)
+  auto tv73 = tv68.expand({dim0, dim1}); // expand tv68
+  auto tv75 = 1.0 / 2048.0; // reciprocal(2048.0)
+  auto tv76 = tv69 * tv75; // mul(tv69, tv75)
+  auto tv77 = 0.000488281; // constant
+  auto tv78 = tv77 * tv73; // mul(tv77, tv73)
+  auto tv79 = tv21 * tv31; // mul(tv21, tv31)
+  auto tv80 = tv78 + tv76; // add(tv78, tv76)
+  auto tv81 = tv79 * tv13; // mul(tv79, tv13)
+  auto tv82 = tv36 + tv80; // add(tv36, tv80)
+  auto tv83 = tv81.sum(0, false); // sum(tv81, {0})
+  auto tv84 = tv13.sum(0, false); // sum(tv13, {0})
+
+  // Expected outputs (cast to BFloat16)
+  auto expected_output0 = tv84.to(at::kBFloat16); // tv87
+  auto expected_output1 = tv83.to(at::kBFloat16); // tv86
+  auto expected_output2 = tv82.to(at::kBFloat16); // tv85
+
+  std::vector<at::Tensor> expected_outputs = {
+      expected_output0, expected_output1, expected_output2};
+
+  testValidate(
+      &fusion_copy, cg_outputs, args, expected_outputs, __LINE__, __FILE__);
+
+  // TODO: Fix auto validation - the fusion runs but testValidate has type
+  // conversion issues testValidate(&fusion_copy, cg_outputs, args,
+  // expected_outputs,
+  // __LINE__, __FILE__);
```

Comment:
errr, the reference implementation looks painful.

What's the issue with the validation? is it type inference that went wrong?